File: netDxf/Tables/ApplicationRegistry.Binding.cs
using System.Collections.Generic;
using System.Linq;
using netDxf.Collections;

namespace netDxf.Tables
{
    public partial class ApplicationRegistry
    {
        private readonly HashSet<XDataDictionary> xdataBindings = new HashSet<XDataDictionary>();
        internal void AttachXData(XDataDictionary dictionary) { this.xdataBindings.Add(dictionary); }
        internal void DetachXData(XDataDictionary dictionary) { this.xdataBindings.Remove(dictionary); }

        /// <summary>Updates internal indexes only after public name observers accept the change.</summary>
        protected override void OnNameChangedEvent(string oldName, string newName)
        {
            this.ValidateBindingRename(newName);
            base.OnNameChangedEvent(oldName, newName);
            // Observers may add/remove owners or XData; validate the current bindings again.
            this.ValidateBindingRename(newName);
            XDataDictionary[] bindings = this.xdataBindings.ToArray();
            if (this.Owner != null) this.Owner.CommitRecordRename(this, newName);
            foreach (XDataDictionary dictionary in bindings) dictionary.CommitApplicationRegistryRename(this, newName);
        }
        private void ValidateBindingRename(string newName)
        {
            if (this.Owner != null) this.Owner.ValidateRecordRename(this, newName);
            foreach (XDataDictionary dictionary in this.xdataBindings)
                dictionary.ValidateApplicationRegistryRename(this, newName);
        }
    }
}

File: javascript/netDxf/Tables/ApplicationRegistry.Binding.js
// Copyright (c) Daniel Carvajal. MIT License; see package LICENSE.
// State shared only with ApplicationRegistry.cs's matching JavaScript partial implementation.
const bindings = new WeakMap();
function state(registry) { let value = bindings.get(registry); if (!value) bindings.set(registry, value = new Set()); return value; }
export function AttachXData(registry, dictionary) { state(registry).add(dictionary); }
export function DetachXData(registry, dictionary) { state(registry).delete(dictionary); }
function ValidateBindingRename(registry, newName) {
  registry.Owner?.ValidateRecordRename(registry, newName);
  for (const dictionary of state(registry)) dictionary.ValidateApplicationRegistryRename(registry, newName);
}
export function OnNameChangedEvent(registry, oldName, newName, notify) {
  ValidateBindingRename(registry, newName); notify(oldName, newName);
  ValidateBindingRename(registry, newName);
  const current = Array.from(state(registry));
  registry.Owner?.CommitRecordRename(registry, newName);
  for (const dictionary of current) dictionary.CommitApplicationRegistryRename(registry, newName);
}
